Add Feature Group object

dromo77 opened the pull request:
## Overview

This PR adds a Featured Row object. The pattern provides a layout for 3 main areas, a heading, content, and and optional cta.

## Screenshots

![Screen Shot 2021-07-13 at 10 13 23 AM](https://user-images.githubusercontent.com/42841342/125497005-dd610c94-c637-4ac9-891a-4a8b5528f129.png)
![Screen Shot 2021-07-13 at 10 19 10 AM](https://user-images.githubusercontent.com/42841342/125497131-d8d9cff3-b9a7-4c7a-aeb1-c21b3aff0ed1.png)

---

- Fixes #1375

tylersticka commented on src/objects/featured-row/featured-row.scss:
```diff
@@ -0,0 +1,42 @@
+  }
+}
+
+.o-featured-row__heading {
+  @media (min-width: breakpoint.$l) {
+    grid-column: 1 / 2;
+    grid-row: 1 / 2;
+  }
+}
```
This is a complete nitpick, but since this class applies to an element that _contains_ a  heading (rather than directly to the heading itself), it may be better to call it something like `header` or `intro`.

tylersticka commented on src/objects/featured-row/featured-row.scss:
```diff
@@ -0,0 +1,42 @@
+  }
+}
+
+.o-featured-row__content {
+  @media (min-width: breakpoint.$l) {
+    grid-column: 2 / 4;
+    grid-row: 1 / 3;
+  }
+}
+
+.o-featured-row__action {
+  @media (min-width: breakpoint.$l) {
+    align-self: end;
+    grid-column: 1 / 2;
+    grid-row: 2 / 3;
+  }
+}
```
Did you consider using `grid-template-areas` to simplify these layout styles?

An example:

```scss

.o-featured-row {
  grid-template-areas:
    'heading' // Each line represents a row
    'content' // Each keyword represents a column
    'action';
  grid-template-rows: repeat(3, minmax(0, auto));

  @media (min-width: breakpoint.$l) {
    // At large breakpoints we update our grid layout, so we also update
    // our areas.
    grid-template-areas:
      'heading content content'
      '. content content' // The dot is just a placeholder to say "leave this empty"
      'action content content';
    grid-template-columns: repeat(3, 1fr);
    grid-template-rows: minmax(0, auto) 1fr minmax(0, auto);
  }
}

// Now all our layout is defined in one place, all that's left is to tell the
// grid which area to put every child element

.o-featured-row__heading {
  grid-area: heading;
}

.o-featured-row__content {
  grid-area: content;
}

.o-featured-row__action {
  grid-area: action;
}
```

tylersticka commented on src/objects/featured-row/featured-row.twig:
```diff
@@ -0,0 +1,15 @@
+    {% block content %}{% endblock %}
+  </div>
+  {% if _action_block is not empty %}
+  <div class="o-featured-row__action">
+    {{_action_block}}
+  </div>
```
Nit: Missing indentation?

dromo77 commented:
Thanks @tylersticka! I made a few changes 🙂 

- Changed `o-featured-row__heading` to `o-featured-row__intro`
- Updated the CSS so it uses `grid-template-areas`
- Fixed the indentation

tylersticka commented on src/objects/featured-row/featured-row.scss:
```diff
@@ -0,0 +1,47 @@
+  @media (min-width: breakpoint.$l) {
+    grid-template-areas:
+      'intro content content'
+      '. content content'
```
I think I made a mistake, because as-is it looks like the `action` is always _below_ the content. I think this might work better:

```suggestion
      'intro content content'
```

tylersticka commented on src/objects/featured-row/featured-row.scss:
```diff
@@ -0,0 +1,47 @@
+}
+
+.o-featured-row__intro {
+  @media (min-width: breakpoint.$l) {
+    grid-area: intro;
+  }
+}
+
+.o-featured-row__content {
+  @media (min-width: breakpoint.$l) {
+    grid-area: content;
+  }
+}
+
+.o-featured-row__action {
+  @media (min-width: breakpoint.$l) {
+    grid-area: action;
+  }
```
I believe the media queries are no longer necessary since the `grid-area` name for these does not change?

tylersticka commented on src/objects/featured-row/featured-row.stories.mdx:
```diff
@@ -0,0 +1,18 @@
+
+<Meta title="Objects/Featured Row" />
+
+# Featured Row
```
Nit: Wondering if "Feature Row" works just as well as the past-tense "Featured"?

Also wondering if "Row" is any more accurate than "Group" or something like that?

tylersticka commented on .changeset/slimy-ears-judge.md:
```diff
@@ -0,0 +1,5 @@
+'@cloudfour/patterns': minor
+---
+
+Add Featured Row object
```
@dromo77 Missed update?

dromo77 replied:
Good catch, thanks!

tylersticka commented on src/objects/feature-group/feature-group.stories.mdx:
```diff
@@ -0,0 +1,18 @@
+import { Story, Canvas, Meta } from '@storybook/addon-docs/blocks';
+import defaultDemo from './demo/demo.twig';
+
+<Meta title="Objects/Feature Group" />
+
+# Feature Group
+
+The Feature Group layout object can be used by individual sections on a page to place content. It is helpful for highlighting a section by giving it a distinct layout. It includes an introduction, main content area, and an optional link.
+
+<Canvas>
+  <Story name="Default">{defaultDemo}</Story>
+</Canvas>
```
@gerardo-rodriguez Could you provide guidance to @dromo77 on how to provide a source code example for this without duplicating her template code?

gerardo-rodriguez replied:
Sure thing! I'll take a closer look this morning. 😉

gerardo-rodriguez replied:
@dromo77 Alrighty, let's look into adding a more helpful source code example. The goal is to arrive at an end result like the following:
<img width="1392" alt="Screen Shot 2021-07-14 at 10 55 16 AM" src="https://user-images.githubusercontent.com/459757/125669814-07727389-fcf8-459c-8175-94a5eb636f9c.png">


We'll follow [a previous PR I merged](https://github.com/cloudfour/cloudfour.com-patterns/pull/1359/files) as an example. 🙂 To set it up, we can leverage the `demo.twig` template you created. This includes:
1. `import` your `demo.twig` file as raw source code
2. Set the imported raw source code as the `docs.source.code` value for the story's `parameters` attribute

Below are the changes you'll need to make. Notice:

- We import raw source code from your template using the non-standard, Webpack-specific`!!raw-loader!` syntax
- We add a helpful comment because it's a non-standard syntax
- We need to disable ESLint for the non-standard syntax `import` line
- We set the `docs.source.code` value to the raw code from your template.

```suggestion
import { Story, Canvas, Meta } from '@storybook/addon-docs/blocks';
// The '!!raw-loader!' syntax is a non-standard, Webpack-specific, syntax.
// See: https://github.com/webpack-contrib/raw-loader#examples
// For now, it seems likely Storybook is pretty tied to Webpack, therefore, we are
// okay with the following Webpack-specific raw loader syntax. It's better to leave
// the ESLint rule enabled globally, and only thoughtfully disable as needed (e.g.
// within a Storybook docs page and not within an actual component).
// This can be revisited in the future if Storybook no longer relies on Webpack.
// eslint-disable-next-line @cloudfour/import/no-webpack-loader-syntax
import defaultDemoSource from '!!raw-loader!./demo/demo.twig';
import defaultDemo from './demo/demo.twig';

<Meta title="Objects/Feature Group" />

# Feature Group

The Feature Group layout object can be used by individual sections on a page to place content. It is helpful for highlighting a section by giving it a distinct layout. It includes an introduction, main content area, and an optional link.

<Canvas>
  <Story
    name="Default"
    parameters={{
      docs: {
        source: {
          code: defaultDemoSource,
        },
      },
    }}
  >
    {defaultDemo}
  </Story>
</Canvas>
```



Feel free to start with the above suggestion. In theory, it _should_ work but if it doesn't let me know and I can help troubleshoot. I have experienced the MDX files to be a bit touchy so don't hesitate to reach out of something breaks. 😉

dromo77 replied:
This is working great, thanks for the helpful explanation @gerardo-rodriguez!

tylersticka reviewed (commented):
@dromo77 I think you may also need to reduce or remove the `grid-row-gap` at larger sizes? I'm noticing that the gap is pushing the action down pretty low:

<img width="1183" alt="Screen Shot 2021-07-14 at 11 30 07 AM" src="https://user-images.githubusercontent.com/69633/125673994-c76e441d-f045-4e6a-8cea-c4d70453af7b.png">

Here it is with `grid-row-gap: 0` added at larger breakpoints:

<img width="1182" alt="Screen Shot 2021-07-14 at 11 30 13 AM" src="https://user-images.githubusercontent.com/69633/125674032-92d2ce65-f530-4826-a8d4-d7734f5beee8.png">

dromo77 commented:
> I'm noticing that the gap is pushing the action down pretty low

@tylersticka I noticed that as well, this seems like a good solution!

tylersticka commented on src/objects/feature-group/feature-group.stories.mdx:
```diff
@@ -0,0 +1,38 @@
+import defaultDemoSource from '!!raw-loader!./demo/demo.twig';
+import defaultDemo from './demo/demo.twig';
+
+<Meta title="Objects/Feature Group" />
```
Since this pattern depends on media queries, you should add `parameters={{ docs: { inlineStories: false } }}` to this `Meta` element. This will cause the stories to display in an iframe in Docs view.

tylersticka reviewed (approved):
🎉
